Issue 37: implemented Lecturers API, added Position entity and migrations, updated Lecturer <-> Department connection

YaroslavDrovolskyi opened the pull request:
**Changes:**

- implemented Lecturers API
- added `Position` entity, migrations, repository, tests
- updated `Lecturer` <-> `Department` connection accoding to new ER-diagram: `DepartmentLecturer` -> `LecturerDepartmentPosition` (updated entity, migrations, repository, tests)

YaroslavDrovolskyi committed:
fix: fix error "Optional<DepartmentResponse> does not have an accessible constructor."

YaroslavDrovolskyi committed:
feat: add application.properties file for test

YaroslavDrovolskyi committed:
feat: impl Position entity, migrations, tests

YaroslavDrovolskyi committed:
feat: change DepartmentLecturer to LecturerDepartmentPosition

updated according to ER diagram.
changed entity, repository, tests and related classes

YaroslavDrovolskyi committed:
feat: implemented LecturerService (NOT TESTED yet)

YaroslavDrovolskyi committed:
feat: implemented LecturerController

YaroslavDrovolskyi committed:
fix: fix "mapper is null" error

also changed "department_positions" field in LecturerDetailsResponse to "lecturer_department_positions"

YaroslavDrovolskyi committed:
fix: fix PERSIST connection Lecturer <-> User

YaroslavDrovolskyi committed:
Merge branch 'development' into issue-37/implement-lecturers-api

YaroslavDrovolskyi committed:
fix: fix minor residue (with DepartmentShortDetailsResponse) after the merge commit

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer/controller/LecturerController.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @GetMapping
+    @PreAuthorize("hasAnyAuthority('ORGANIZER', 'SYSADMIN', 'ADMIN', 'LECTURER', 'STUDENT')")
```
You can just use "USER" role for this.

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer/controller/LecturerController.java:
```diff
@@ -0,0 +1,92 @@
+
+    @PostMapping("/{id}/department-positions")
+    @PreAuthorize("hasAnyAuthority('ORGANIZER', 'SYSADMIN', 'ADMIN')")
+    public ResponseEntity<LecturerDetailsResponse> addDepartmentPosition(
```
I think, that you forgot about deleteDepartmentPosition endpoint

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer/model/LecturerDetailsResponse.java:
```diff
@@ -0,0 +1,18 @@
+
+        UserDetailsResponse user,
+
+        List<DepartmentPositionDto> lecturerDepartmentPositions,
```
maybe just rename to departmentPositions?

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer/model/LecturerReplaceAcademicTitleRequest.java:
```diff
@@ -0,0 +1,8 @@
+package education.kub.backend.ce.domain.lecturer.model;
+
+public record LecturerReplaceAcademicTitleRequest(
+        Long oldAcademicTitleId,
```
I think we can just use academicTitleId, because it will be edit request. to lecturers/{id_l}/academic-titles/{id}

?

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer/service/LecturerService.java:
```diff
@@ -0,0 +1,154 @@
+        return lecturerMapper.toDetailsResponseList(lecturerRepository.findAll());
+    }
+
+    public LecturerDetailsResponse getLecturerFullById(Long id) {
```
why lecturerFull? I think we return every time full lecturer. just rename to getLecturerById.

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer/service/LecturerService.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+
+    public LecturerDetailsResponse addDepartmentPosition(Long lecturerId,
```
i think we need also deleteDepartmentPosition

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer/service/LecturerService.java:
```diff
@@ -0,0 +1,154 @@
+
+    // use case: when lecturer updates his academic title from candidate to doctor os science
+    @Transactional
+    public LecturerDetailsResponse replaceAcademicTitle(Long lecturerId,
```
Overthinking. Why just not create update method by id?

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer_department_position/model/LecturerDepartmentPositionCreateRequest.java:
```diff
@@ -0,0 +1,19 @@
+        Long positionId,
+
+        @NotNull(message = "status can't be null")
+        LecturerDepartmentPositionEntity.Status status
```
do we need add position status for create?

PavloKilko commented on kub-backend-ce/src/main/java/education/kub/backend/ce/domain/lecturer_department_position/service/LecturerDepartmentPositionService.java:
```diff
@@ -0,0 +1,59 @@
+        return lecturerDepartmentPositionMapper.toDto(lecturerDepartmentPosition);
+    }
+
+    public void deleteLecturerDepartmentPosition(Long lecturerId, Long departmentId){
```
Where is an update method?

YaroslavDrovolskyi replied:
In department position API?

YaroslavDrovolskyi committed:
fix: fix Lecturer API

Fix Lecturer API according to Pavlo Kilko's comments. Changes include new enpoins for managing lecturer's departments positions and academic titles

YaroslavDrovolskyi committed:
fix: updated Lecturer API according to docs

- deleted updateAcademicTitle endpoint
- made only ADMIN (not ORGANIZER, SYSADMIN) responsible for Lecturers management
- fixed addAcademicTitle endpoint having body (path argument is instead of body, now)

PavloKilko committed:
refactor: change names for service-level objects to Dto

PavloKilko reviewed (approved):
lgtm

OleksandrTomas reviewed (approved):
lgtm

PavloKilko reviewed (changes requested):
Add API paths for /positions /academic-titles.

YaroslavDrovolskyi committed:
feat: add /positions and /academic-titles API
